src/ProjectOrigin.Vault/CommandHandlers/CheckForWithdrawnCertificatesCommandHandler.cs

```diff
@@ -0,0 +1,108 @@
+{
+}
+
+public class CheckForWithdrawnCertificatesCommandHandler : IConsumer<CheckForWithdrawnCertificatesCommand>
+{
+    private readonly IUnitOfWork _unitOfWork;
+    private readonly IHttpClientFactory _httpClientFactory;
+    private readonly ILogger<CheckForWithdrawnCertificatesCommandHandler> _logger;
+    private readonly NetworkOptions _networkOptions;
+    public CheckForWithdrawnCertificatesCommandHandler(IUnitOfWork unitOfWork,
+        IOptions<NetworkOptions> networkOptions,
+        IHttpClientFactory httpClientFactory,
+        ILogger<CheckForWithdrawnCertificatesCommandHandler> logger)
+    {
+        _unitOfWork = unitOfWork;
+        _httpClientFactory = httpClientFactory;
+        _logger = logger;
+        _networkOptions = networkOptions.Value;
+    }
+
+    public async Task Consume(ConsumeContext<CheckForWithdrawnCertificatesCommand> context)
+    {
+        var stamps = _networkOptions.Stamps;
+
+        foreach (var stamp in stamps)
+        {
+            var cursors = await _unitOfWork.WithdrawnCursorRepository.GetWithdrawnCursors();
+            var matchingCursor = cursors.FirstOrDefault(x => x.StampName == stamp.Key) ?? new WithdrawnCursor
+            {
+                StampName = stamp.Key,
+                SyncPosition = 0,
+                LastSyncDate = DateTimeOffset.UtcNow
+            };
+
+            var client = _httpClientFactory.CreateClient();
+            var response = (await client.GetFromJsonAsync<ResultList<WithdrawnCertificateDto, PageInfo>>(stamp.Value.Url + $"/v1/certificates/withdrawn?lastWithdrawnId={matchingCursor.SyncPosition}"))!;
+
+            if (!response.Result.Any())
+            {
+                _logger.LogInformation("No withdrawn certificates found for {StampName}", stamp.Key);
+                continue;
+            }
+
+            _logger.LogInformation("Found {Count} withdrawn certificates for {StampName}", response.Result.Count(), stamp.Key);
+            foreach (var withdrawnCertificate in response.Result)
+            {
+                await _unitOfWork.CertificateRepository.WithdrawCertificate(withdrawnCertificate.RegistryName, withdrawnCertificate.CertificateId);
+                var claimedSlices = await _unitOfWork.CertificateRepository.GetClaimedSlicesOfCertificate(withdrawnCertificate.RegistryName, withdrawnCertificate.CertificateId);
+                foreach (var claimedSlice in claimedSlices)
+                {
+                    //Unclaim (which is next task)
+                }
+            }
+
+            matchingCursor.SyncPosition = response.Result.Max(x => x.Id);
+            matchingCursor.LastSyncDate = DateTimeOffset.UtcNow;
+            await UpdateWithdrawnCursor(matchingCursor);
+            _unitOfWork.Commit();
+        }
+    }
+
+    private async Task UpdateWithdrawnCursor(WithdrawnCursor updatedCursor)
+    {
+        var cursors = await _unitOfWork.WithdrawnCursorRepository.GetWithdrawnCursors();
```

Comment:
This could be done with an Upsert statement in the database, so we don't need to check if it exists prior to creating it.

Reply:
done